File: .github/scripts/analyze-results.py
#!/usr/bin/env python3
"""
Analyze Claude test results and generate summary
"""

import json
import glob
import os
import sys
from pathlib import Path
from datetime import datetime
from collections import defaultdict

def load_results(results_dir):
    """Load all test result JSON files"""
    results = []
    pattern = os.path.join(results_dir, '**/*.json')
    
    for filepath in glob.glob(pattern, recursive=True):
        try:
            with open(filepath, 'r') as f:
                data = json.load(f)
                results.append(data)
        except json.JSONDecodeError:
            print(f"Warning: Could not parse {filepath}", file=sys.stderr)
        except Exception as e:
            print(f"Warning: Error loading {filepath}: {e}", file=sys.stderr)
    
    return results

def calculate_stats(results):
    """Calculate statistics across all results"""
    if not results:
        return {}
    
    stats = {
        'by_test': defaultdict(lambda: {
            'scores': [],
            'passed_count': 0,
            'failed_count': 0
        }),
        'overall': {
            'total_results': len(results),
            'average_score': 0,
            'average_pass_rate': 0
        }
    }
    
    total_score = 0
    total_pass_rate = 0
    
    for result in results:
        total_score += result.get('averageScore', 0)
        total_pass_rate += result.get('passRate', 0)
        
        for test in result.get('tests', []):
            test_id = test.get('testId', 'unknown')
            test_stats = stats['by_test'][test_id]
            test_stats['scores'].append(test.get('score', 0))
            if test.get('passed', False):
                test_stats['passed_count'] += 1
            else:
                test_stats['failed_count'] += 1
    
    stats['overall']['average_score'] = round(total_score / len(results)) if results else 0
    stats['overall']['average_pass_rate'] = round(total_pass_rate / len(results)) if results else 0
    
    return stats

def generate_summary_markdown(stats, output_file):
    """Generate markdown summary report"""
    
    md = []
    md.append("# Claude Code Test Results")
    md.append("")
    md.append(f"**Generated**: {datetime.now().strftime('%Y-%m-%d %H:%M:%S UTC')}")
    md.append("")
    
    md.append("## Overall Summary")
    md.append("")
    md.append(f"- **Total Test Runs**: {stats['overall']['total_results']}")
    md.append(f"- **Average Score**: {stats['overall']['average_score']}/100")
    md.append(f"- **Average Pass Rate**: {stats['overall']['average_pass_rate']}%")
    md.append("")
    
    if stats['overall']['average_pass_rate'] >= 95:
        md.append("✅ **Status**: EXCELLENT - Producing consistent high-quality code")
    elif stats['overall']['average_pass_rate'] >= 85:
        md.append("⚠️  **Status**: GOOD - Minor improvements needed")
    elif stats['overall']['average_pass_rate'] >= 75:
        md.append("⚠️  **Status**: FAIR - Significant quality gaps")
    else:
        md.append("❌ **Status**: POOR - Major quality issues")
    md.append("")
    
    md.append("## Results by Test")
    md.append("")
    md.append("| Test | Passed | Failed | Avg Score |")
    md.append("|------|--------|--------|-----------|")
    
    for test_id, test_stats in sorted(stats['by_test'].items()):
        avg_score = round(sum(test_stats['scores']) / len(test_stats['scores'])) if test_stats['scores'] else 0
        md.append(f"| {test_id} | {test_stats['passed_count']} | {test_stats['failed_count']} | {avg_score}/100 |")
    
    md.append("")
    
    md.append("## Recommendations")
    md.append("")
    
    if stats['overall']['average_pass_rate'] < 95:
        md.append("### Actions Required")
        md.append("")
        
        problematic_tests = []
        for test_id, test_stats in stats['by_test'].items():
            total = test_stats['passed_count'] + test_stats['failed_count']
            failure_rate = (test_stats['failed_count'] / total) * 100 if total > 0 else 0
            if failure_rate > 20:
                problematic_tests.append((test_id, failure_rate))
        
        if problematic_tests:
            md.append("**Tests with high failure rates:**")
            for test_id, failure_rate in sorted(problematic_tests, key=lambda x: -x[1]):
                md.append(f"- **{test_id}**: {round(failure_rate)}% failure rate")
            md.append("")
            md.append("Review these test scenarios and strengthen related rule files.")
        
        md.append("")
        md.append("**Recommended fixes:**")
        md.append("1. Add more explicit ALWAYS/NEVER directives to affected framework files")
        md.append("2. Include more 'Common AI Mistakes' examples")
        md.append("3. Re-run tests after applying fixes")
    else:
        md.append("✅ All tests performing excellently! No immediate actions required.")
        md.append("")
        md.append("**Maintenance:**")
        md.append("- Continue monitoring test runs")
        md.append("- Review any new failures promptly")
        md.append("- Keep rule files updated with new patterns")
    
    md.append("")
    md.append("---")
    md.append("*For detailed results, check individual test artifacts.*")
    
    with open(output_file, 'w') as f:
        f.write('\n'.join(md))
    
    print(f"Summary written to {output_file}")

def main():
    import argparse
    
    parser = argparse.ArgumentParser(description='Analyze Claude test results')
    parser.add_argument('--results-dir', required=True, help='Directory containing test results')
    parser.add_argument('--output', required=True, help='Output markdown file')
    
    args = parser.parse_args()
    
    print(f"Loading results from {args.results_dir}...")
    results = load_results(args.results_dir)
    print(f"Loaded {len(results)} result files")
    
    if not results:
        print("ERROR: No results found", file=sys.stderr)
        sys.exit(1)
    
    print("Calculating statistics...")
    stats = calculate_stats(results)
    
    print("Generating summary...")
    generate_summary_markdown(stats, args.output)
    
    print("Done!")

if __name__ == '__main__':
    main()

File: .github/scripts/analyze-results.js
#!/usr/bin/env node

/**
 * Claude Code Results Analyzer
 * 
 * Analyzes test results from Claude to ensure rule quality
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const resultsDir = process.argv[2] || './all-results';

function loadResults(dir) {
  const results = [];
  
  if (!fs.existsSync(dir)) {
    console.error(`Results directory not found: ${dir}`);
    return results;
  }
  
  const entries = fs.readdirSync(dir);
  
  for (const entry of entries) {
    const entryPath = path.join(dir, entry);
    const stat = fs.statSync(entryPath);
    
    if (stat.isDirectory()) {
      const files = fs.readdirSync(entryPath).filter(f => f.endsWith('.json'));
      for (const file of files) {
        const filePath = path.join(entryPath, file);
        try {
          const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
          results.push(data);
        } catch (error) {
          console.error(`Error loading ${filePath}:`, error.message);
        }
      }
    } else if (entry.endsWith('.json')) {
      try {
        const data = JSON.parse(fs.readFileSync(entryPath, 'utf8'));
        results.push(data);
      } catch (error) {
        console.error(`Error loading ${entryPath}:`, error.message);
      }
    }
  }
  
  return results;
}

function analyzeResults(results) {
  if (results.length === 0) {
    return { error: 'No results to analyze' };
  }
  
  const testGroups = {};
  
  for (const result of results) {
    for (const test of result.tests) {
      if (!testGroups[test.testId]) {
        testGroups[test.testId] = [];
      }
      testGroups[test.testId].push({
        score: test.score,
        passed: test.passed,
        validation: test.validation
      });
    }
  }
  
  const testAnalysis = {};
  
  for (const [testId, testResults] of Object.entries(testGroups)) {
    const scores = testResults.map(r => r.score);
    const avgScore = scores.reduce((a, b) => a + b, 0) / scores.length;
    const passRate = (testResults.filter(r => r.passed).length / testResults.length) * 100;
    
    testAnalysis[testId] = {
      totalRuns: testResults.length,
      avgScore: Math.round(avgScore),
      passRate: Math.round(passRate),
      results: testResults
    };
  }
  
  const allScores = results.flatMap(r => r.tests.map(t => t.score));
  const overallAvg = Math.round(allScores.reduce((a, b) => a + b, 0) / allScores.length);
  
  return {
    overallAverage: overallAvg,
    totalTests: Object.keys(testGroups).length,
    totalRuns: results.length,
    testAnalysis,
    passing: overallAvg >= 90
  };
}

function generateReport(analysis) {
  let report = '# Claude Code Test Results\n\n';
  
  if (analysis.error) {
    report += `## Error\n\n${analysis.error}\n`;
    return report;
  }
  
  report += '## Summary\n\n';
  report += `- **Overall Average Score**: ${analysis.overallAverage}/100\n`;
  report += `- **Total Tests**: ${analysis.totalTests}\n`;
  report += `- **Total Runs**: ${analysis.totalRuns}\n`;
  report += `- **Status**: ${analysis.passing ? '✅ PASS' : '⚠️ NEEDS ATTENTION'}\n`;
  report += '\n';
  
  report += '## Test-by-Test Analysis\n\n';
  
  for (const [testId, test] of Object.entries(analysis.testAnalysis)) {
    const passIcon = test.passRate === 100 ? '✅' : test.passRate >= 80 ? '⚠️' : '❌';
    
    report += `### ${passIcon} ${testId}\n\n`;
    report += `- **Average Score**: ${test.avgScore}/100\n`;
    report += `- **Pass Rate**: ${test.passRate}%\n`;
    report += '\n';
  }
  
  report += '## Recommendations\n\n';
  
  if (analysis.passing) {
    report += '✅ **All tests passing!** Rules are producing high-quality code with Claude.\n\n';
  } else {
    report += '⚠️ **Action items**:\n\n';
    
    const failingTests = Object.entries(analysis.testAnalysis)
      .filter(([_, test]) => test.passRate < 100)
      .map(([id, _]) => id);
    
    if (failingTests.length > 0) {
      report += `1. **Improve rules** for failing tests: ${failingTests.join(', ')}\n`;
      report += '   - Add more explicit `> **ALWAYS**` / `> **NEVER**` directives\n';
      report += '   - Strengthen pattern descriptions in rules\n';
      report += '   - Add more code examples with correct/incorrect patterns\n\n';
    }
    
    if (analysis.overallAverage < 90) {
      report += '2. **Improve overall scores**:\n';
      report += '   - Review failed pattern matches in test results\n';
      report += '   - Ensure rules are clear and unambiguous\n';
      report += '   - Add AI Self-Check sections to catch common mistakes\n\n';
    }
  }
  
  report += '## Success Criteria\n\n';
  report += 'Rules should produce consistent, high-quality code:\n';
  report += '- ✅ Consistent code structure and architecture patterns\n';
  report += '- ✅ Same security practices applied (OWASP Top 10, input validation)\n';
  report += '- ✅ Same testing approaches (unit, integration, mocking)\n';
  report += '- ✅ Same CI/CD best practices (version detection, caching)\n\n';
  
  const criteriaCheck = analysis.passing;
  report += '**Current Status**: ';
  report += criteriaCheck 
    ? '✅ **MEETING CRITERIA**\n'
    : '⚠️ **NOT YET MEETING CRITERIA** - See recommendations above\n';
  
  return report;
}

function main() {
  console.log('Loading test results...');
  const results = loadResults(resultsDir);
  
  console.log(`Found ${results.length} result files`);
  
  if (results.length === 0) {
    console.log('No results to analyze. This is expected if no tests ran (e.g., missing API key).');
    const report = generateReport({ error: 'No results found. Tests may not have run.' });
    fs.writeFileSync('analysis-report.md', report);
    process.exit(0);
  }
  
  console.log('Analyzing results...');
  const analysis = analyzeResults(results);
  
  console.log('Generating report...');
  const report = generateReport(analysis);
  
  fs.writeFileSync('analysis-report.md', report);
  console.log('Report saved to analysis-report.md');
  
  console.log('\n=== Summary ===');
  console.log(`Overall Average: ${analysis.overallAverage}/100`);
  console.log(`Status: ${analysis.passing ? 'PASS' : 'FAIL'}`);
  
  if (!analysis.passing) {
    console.error('\nERROR: Not meeting success criteria');
    process.exit(1);
  }
}

main();
